node/bft/src/helpers/telemetry.rs

```diff
@@ -0,0 +1,368 @@
+
+    /// Insert a subdag to the telemetry tracker.
+    /// Note: This currently assumes the subdag is fully formed and included in the block.
+    pub fn insert_subdag(&self, subdag: &Subdag<N>) {
+        // Garbage collect the old certificates.
+        let next_gc_round = subdag.anchor_round().saturating_sub(BatchHeader::<N>::MAX_GC_ROUNDS as u64);
+        self.garbage_collect_certificates(next_gc_round);
+
+        // Insert the subdag certificates.
+        cfg_iter!(subdag).for_each(|(_round, certificates)| {
+            cfg_iter!(certificates).for_each(|certificate| {
+                // TODO (raychu86): Can be greatly optimized by doing a one-shot update instead of individual certificates.
+                self.insert_certificate(certificate);
+            })
+        });
+
+        // Calculate the participation scores.
+        self.update_participation_scores();
+    }
+
+    /// Insert a certificate to the telemetry tracker.
+    pub fn insert_certificate(&self, certificate: &BatchCertificate<N>) {
+        // Acquire the lock for `tracked_certificates`.
+        let mut tracked_certificates = self.tracked_certificates.write();
+
+        // Retrieve the certificate round, author, and ID.
+        let certificate_round = certificate.round();
+        let certificate_author = certificate.author();
+        let certificate_id = certificate.id();
+
+        // If the certificate already exists in the tracker, then return early.
+        if tracked_certificates.get(&certificate_round).map_or(false, |certs| certs.contains(&certificate_id)) {
+            return;
+        }
+
+        // Insert the certificate ID.
+        tracked_certificates.entry(certificate_round).or_default().insert(certificate_id);
+
+        // Acquire the lock for `validator_signatures`
+        let mut validator_signatures = self.validator_signatures.write();
+
+        // Insert the certificate author and signers.
+        for address in
+            [certificate_author].into_iter().chain(certificate.signatures().map(|signature| signature.to_address()))
+        {
+            validator_signatures
+                .entry(address)
+                .or_default()
+                .entry(certificate_round)
+                .and_modify(|count| *count += 1)
+                .or_insert(1);
+        }
+
+        // Acquire the lock for `validator_certificates`.
+        let mut validator_certificates = self.validator_certificates.write();
+
+        // Insert the certificate
+        validator_certificates.entry(certificate_author).or_default().insert(certificate_round);
+    }
+
+    /// Calculate and update the participation scores for each validator.
+    pub fn update_participation_scores(&self) {
+        // Fetch the certificates and signatures.
+        let tracked_certificates = self.tracked_certificates.read();
+        let validator_signatures = self.validator_signatures.read();
+        let validator_certificates = self.validator_certificates.read();
```

Comment:
these 3 collections appear to always change together; it would be more performant if they were all under a single `RwLock`

Reply:
The intention in the future is to add additional metrics that could be independent. Which is why I kept the trackers as individual Arcs.

Reply:
we can always put those 3 under a single lock in case we need extra performance